File: NextApp/Android/core/data/src/main/kotlin/jp/komehyappyo/member/next/core/data/FirebaseRestPostRepository.kt
package jp.komehyappyo.member.next.core.data

import jp.komehyappyo.member.next.core.model.AdminReply
import jp.komehyappyo.member.next.core.model.MemberPost
import jp.komehyappyo.member.next.core.model.PostAttachment
import jp.komehyappyo.member.next.core.model.PublicPost
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONArray
import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URL
import java.time.Instant
import java.util.UUID

interface PostRepository {
    suspend fun publicPosts(): Result<List<PublicPost>>
    suspend fun memberPosts(
        communityId: String,
        userId: String,
        idToken: String,
    ): Result<List<MemberPost>>
    suspend fun replies(
        communityId: String,
        postId: String,
        idToken: String,
    ): Result<List<AdminReply>>
    suspend fun createMemberPost(
        communityId: String,
        userId: String,
        memberName: String,
        title: String,
        body: String,
        idToken: String,
    ): Result<Unit>
    suspend fun updateMemberPost(
        communityId: String,
        postId: String,
        title: String,
        body: String,
        idToken: String,
    ): Result<Unit>
    suspend fun deleteMemberPost(
        communityId: String,
        postId: String,
        idToken: String,
    ): Result<Unit>
    suspend fun markReplyRead(
        communityId: String,
        postId: String,
        idToken: String,
    ): Result<Unit>
}

class FirebaseRestPostRepository(
    private val projectId: String,
) : PostRepository {
    override suspend fun publicPosts(): Result<List<PublicPost>> = runCatching {
        collectionGroup("publicPosts", null, publicOnly = true)
            .mapNotNull(::parsePublicPost)
            .filter { it.body.isNotBlank() || it.title.isNotBlank() }
            .sortedByDescending { it.createdAt.orEmpty() }
    }

    override suspend fun memberPosts(
        communityId: String,
        userId: String,
        idToken: String,
    ): Result<List<MemberPost>> = runCatching {
        listDocuments("organizations/$communityId/memberPosts", idToken)
            .mapNotNull { parseMemberPost(communityId, it) }
            .filter { it.authorUserId == userId }
            .sortedByDescending { it.updatedAt ?: it.createdAt.orEmpty() }
    }

    override suspend fun replies(
        communityId: String,
        postId: String,
        idToken: String,
    ): Result<List<AdminReply>> = runCatching {
        listDocuments("organizations/$communityId/memberPosts/$postId/replies", idToken)
            .mapNotNull { parseReply(postId, it) }
            .sortedBy { it.createdAt.orEmpty() }
    }

    override suspend fun createMemberPost(
        communityId: String,
        userId: String,
        memberName: String,
        title: String,
        body: String,
        idToken: String,
    ): Result<Unit> = runCatching {
        val now = Instant.now().toString()
        val fields = JSONObject()
            .put("organizationId", stringValue(communityId))
            .put("memberUid", stringValue(userId))
            .put("memberName", stringValue(memberName))
            .put("title", stringValue(title.trim()))
            .put("body", stringValue(body.trim()))
            .put("imageURLs", arrayValue(emptyList()))
            .put("status", stringValue("submitted"))
            .put("adminReply", stringValue(""))
            .put("memberHasReadReply", booleanValue(true))
            .put("createdAt", timestampValue(now))
            .put("updatedAt", timestampValue(now))
        request(
            "documents/organizations/$communityId/memberPosts/${UUID.randomUUID()}",
            "PATCH",
            idToken,
            JSONObject().put("fields", fields),
        )
        Unit
    }

    override suspend fun updateMemberPost(
        communityId: String,
        postId: String,
        title: String,
        body: String,
        idToken: String,
    ): Result<Unit> = runCatching {
        val fields = JSONObject()
            .put("title", stringValue(title.trim()))
            .put("body", stringValue(body.trim()))
            .put("updatedAt", timestampValue(Instant.now().toString()))
        request(
            "documents/organizations/$communityId/memberPosts/$postId" +
                "?updateMask.fieldPaths=title&updateMask.fieldPaths=body" +
                "&updateMask.fieldPaths=updatedAt",
            "PATCH",
            idToken,
            JSONObject().put("fields", fields),
        )
        Unit
    }

    override suspend fun deleteMemberPost(
        communityId: String,
        postId: String,
        idToken: String,
    ): Result<Unit> = runCatching {
        request(
            "documents/organizations/$communityId/memberPosts/$postId",
            "DELETE",
            idToken,
            null,
        )
        Unit
    }

    override suspend fun markReplyRead(
        communityId: String,
        postId: String,
        idToken: String,
    ): Result<Unit> = runCatching {
        request(
            "documents/organizations/$communityId/memberPosts/$postId" +
                "?updateMask.fieldPaths=memberHasReadReply",
            "PATCH",
            idToken,
            JSONObject().put(
                "fields",
                JSONObject().put("memberHasReadReply", booleanValue(true)),
            ),
        )
        Unit
    }

    private suspend fun listDocuments(path: String, idToken: String): List<JSONObject> {
        val result = request("documents/$path?pageSize=1000", "GET", idToken, null) as JSONObject
        val values = result.optJSONArray("documents") ?: return emptyList()
        return buildList {
            for (index in 0 until values.length()) {
                values.optJSONObject(index)?.let(::add)
            }
        }
    }

    private suspend fun collectionGroup(
        collection: String,
        idToken: String?,
        publicOnly: Boolean = false,
    ): List<JSONObject> {
        val structuredQuery = JSONObject().put(
            "from",
            JSONArray().put(
                JSONObject()
                    .put("collectionId", collection)
                    .put("allDescendants", true),
            ),
        )
        if (publicOnly) {
            structuredQuery.put(
                "where",
                JSONObject().put(
                    "fieldFilter",
                    JSONObject()
                        .put("field", JSONObject().put("fieldPath", "isPublic"))
                        .put("op", "EQUAL")
                        .put("value", booleanValue(true)),
                ),
            )
        }
        val query = JSONObject().put("structuredQuery", structuredQuery)
        val rows = request("documents:runQuery", "POST", idToken, query) as JSONArray
        return buildList {
            for (index in 0 until rows.length()) {
                rows.optJSONObject(index)?.optJSONObject("document")?.let(::add)
            }
        }
    }

    internal fun parseMemberPost(communityId: String, document: JSONObject): MemberPost? {
        val fields = document.optJSONObject("fields") ?: return null
        val id = document.optString("name").substringAfterLast("/")
        return MemberPost(
            id = id,
            communityId = string(fields, "organizationId") ?: communityId,
            authorUserId = string(fields, "memberUid")
                ?: string(fields, "authorUserId") ?: return null,
            authorName = string(fields, "memberName")
                ?: string(fields, "authorName") ?: "会員",
            title = string(fields, "title").orEmpty(),
            body = string(fields, "body").orEmpty(),
            attachments = attachments(fields),
            status = string(fields, "status") ?: "submitted",
            adminReply = string(fields, "adminReply").orEmpty(),
            memberHasReadReply = boolean(fields, "memberHasReadReply") ?: true,
            createdAt = timestamp(fields, "createdAt"),
            updatedAt = timestamp(fields, "updatedAt"),
        )
    }

    internal fun parsePublicPost(document: JSONObject): PublicPost? {
        val fields = document.optJSONObject("fields") ?: return null
        if (boolean(fields, "isPublic") != true) return null
        return PublicPost(
            id = document.optString("name").substringAfterLast("/"),
            authorUserId = string(fields, "authorUserId")
                ?: string(fields, "memberUid").orEmpty(),
            authorName = string(fields, "authorName")
                ?: string(fields, "memberName") ?: "投稿者",
            title = string(fields, "title").orEmpty(),
            body = string(fields, "body").orEmpty(),
            categoryId = string(fields, "categoryId") ?: string(fields, "category"),
            attachments = attachments(fields),
            createdAt = timestamp(fields, "createdAt"),
        )
    }

    private fun parseReply(postId: String, document: JSONObject): AdminReply? {
        val fields = document.optJSONObject("fields") ?: return null
        return AdminReply(
            id = document.optString("name").substringAfterLast("/"),
            postId = postId,
            adminUserId = string(fields, "createdBy")
                ?: string(fields, "adminUserId").orEmpty(),
            adminName = string(fields, "createdByName")
                ?: string(fields, "adminName") ?: "管理者",
            body = string(fields, "body") ?: return null,
            createdAt = timestamp(fields, "createdAt"),
        )
    }

    private fun attachments(fields: JSONObject): List<PostAttachment> {
        val result = mutableListOf<PostAttachment>()
        val values = fields.optJSONObject("attachments")
            ?.optJSONObject("arrayValue")?.optJSONArray("values")
        if (values != null) {
            for (index in 0 until values.length()) {
                val entry = values.optJSONObject(index)
                    ?.optJSONObject("mapValue")?.optJSONObject("fields") ?: continue
                val url = string(entry, "url") ?: continue
                result += PostAttachment(
                    type = string(entry, "type") ?: "url",
                    name = string(entry, "name") ?: "添付ファイル",
                    url = url,
                )
            }
        }
        stringArray(fields, "imageURLs").forEachIndexed { index, url ->
            result += PostAttachment("image", "画像 ${index + 1}", url)
        }
        string(fields, "pdfURL")?.let { result += PostAttachment("pdf", "PDF", it) }
        return result.distinctBy(PostAttachment::url)
    }

    private suspend fun request(
        path: String,
        method: String,
        idToken: String?,
        body: JSONObject?,
    ): Any = withContext(Dispatchers.IO) {
        val connection = (
            URL(
                "https://firestore.googleapis.com/v1/projects/$projectId/" +
                    "databases/(default)/$path",
            ).openConnection() as HttpURLConnection
            ).apply {
            requestMethod = method
            connectTimeout = 15_000
            readTimeout = 15_000
            idToken?.let { setRequestProperty("Authorization", "Bearer $it") }
            if (body != null) {
                doOutput = true
                setRequestProperty("Content-Type", "application/json; charset=UTF-8")
            }
        }
        try {
            body?.let {
                connection.outputStream.use { stream ->
                    stream.write(it.toString().toByteArray(Charsets.UTF_8))
                }
            }
            val status = connection.responseCode
            if (status !in 200..299) {
                throw IllegalStateException(
                    if (status == 403) {
                        "投稿を操作する権限がありません。参加承認後に再度お試しください。"
                    } else {
                        "投稿を処理できませんでした。時間をおいて再度お試しください。"
                    },
                )
            }
            if (status == HttpURLConnection.HTTP_NO_CONTENT) return@withContext JSONObject()
            val text = connection.inputStream.bufferedReader().use { it.readText() }
            if (text.isBlank()) JSONObject()
            else if (text.trimStart().startsWith("[")) JSONArray(text) else JSONObject(text)
        } finally {
            connection.disconnect()
        }
    }

    private fun stringValue(value: String) = JSONObject().put("stringValue", value)
    private fun booleanValue(value: Boolean) = JSONObject().put("booleanValue", value)
    private fun timestampValue(value: String) = JSONObject().put("timestampValue", value)
    private fun arrayValue(values: List<String>) = JSONObject().put(
        "arrayValue",
        JSONObject().put(
            "values",
            JSONArray().apply { values.forEach { put(stringValue(it)) } },
        ),
    )
    private fun string(fields: JSONObject, key: String): String? =
        fields.optJSONObject(key)?.optString("stringValue")?.takeIf(String::isNotEmpty)
    private fun boolean(fields: JSONObject, key: String): Boolean? =
        fields.optJSONObject(key)?.takeIf { it.has("booleanValue") }?.optBoolean("booleanValue")
    private fun timestamp(fields: JSONObject, key: String): String? =
        fields.optJSONObject(key)?.optString("timestampValue")?.takeIf(String::isNotEmpty)
    private fun stringArray(fields: JSONObject, key: String): List<String> {
        val values = fields.optJSONObject(key)
            ?.optJSONObject("arrayValue")?.optJSONArray("values") ?: return emptyList()
        return buildList {
            for (index in 0 until values.length()) {
                values.optJSONObject(index)?.optString("stringValue")
                    ?.takeIf(String::isNotEmpty)?.let(::add)
            }
        }
    }
}

File: NextApp/iOS/OrgPortalNext/Core/DataLayer/FirebaseRESTPostRepository.swift
import Foundation
import Model

public protocol PostRepository: Sendable {
    func publicPosts() async throws -> [PublicPost]
    func memberPosts(
        communityId: String,
        userId: String,
        idToken: String
    ) async throws -> [MemberPost]
    func replies(
        communityId: String,
        postId: String,
        idToken: String
    ) async throws -> [AdminReply]
    func createMemberPost(
        communityId: String,
        userId: String,
        authorName: String,
        title: String,
        body: String,
        idToken: String
    ) async throws
    func updateMemberPost(
        communityId: String,
        postId: String,
        title: String,
        body: String,
        idToken: String
    ) async throws
    func deleteMemberPost(
        communityId: String,
        postId: String,
        idToken: String
    ) async throws
    func markReplyRead(
        communityId: String,
        postId: String,
        idToken: String
    ) async throws
}

public struct FirebaseRESTPostRepository: PostRepository {
    private let projectId: String
    private let session: URLSession

    public init(projectId: String, session: URLSession = .shared) {
        self.projectId = projectId
        self.session = session
    }

    public func publicPosts() async throws -> [PublicPost] {
        let rows = try await request(
            path: "documents:runQuery",
            method: "POST",
            body: [
                "structuredQuery": [
                    "from": [["collectionId": "publicPosts", "allDescendants": true]],
                    "where": [
                        "fieldFilter": [
                            "field": ["fieldPath": "isPublic"],
                            "op": "EQUAL",
                            "value": boolValue(true)
                        ]
                    ]
                ]
            ]
        ) as? [[String: Any]] ?? []
        return rows.compactMap { row in
            guard let document = row["document"] as? [String: Any] else { return nil }
            return parsePublicPost(document)
        }
        .sorted { ($0.createdAt ?? .distantPast) > ($1.createdAt ?? .distantPast) }
    }

    public func memberPosts(
        communityId: String,
        userId: String,
        idToken: String
    ) async throws -> [MemberPost] {
        let result = try await request(
            path: "documents/organizations/\(communityId)/memberPosts?pageSize=1000",
            method: "GET",
            idToken: idToken
        ) as? [String: Any]
        return (result?["documents"] as? [[String: Any]] ?? [])
            .compactMap { parseMemberPost($0, communityId: communityId) }
            .filter { $0.authorUserId == userId }
            .sorted {
                ($0.updatedAt ?? $0.createdAt ?? .distantPast)
                    > ($1.updatedAt ?? $1.createdAt ?? .distantPast)
            }
    }

    public func replies(
        communityId: String,
        postId: String,
        idToken: String
    ) async throws -> [AdminReply] {
        let result = try await request(
            path: "documents/organizations/\(communityId)/memberPosts/\(postId)"
                + "/replies?pageSize=1000",
            method: "GET",
            idToken: idToken
        ) as? [String: Any]
        return (result?["documents"] as? [[String: Any]] ?? []).compactMap {
            guard let fields = $0["fields"] as? [String: Any],
                  let name = $0["name"] as? String,
                  let id = name.split(separator: "/").last.map(String.init),
                  let body = string(fields, "body") else { return nil }
            return AdminReply(
                id: id,
                postId: postId,
                adminUserId: string(fields, "createdBy") ?? "",
                adminName: string(fields, "createdByName"),
                body: body,
                createdAt: timestamp(fields, "createdAt")
            )
        }
        .sorted { ($0.createdAt ?? .distantPast) < ($1.createdAt ?? .distantPast) }
    }

    public func createMemberPost(
        communityId: String,
        userId: String,
        authorName: String,
        title: String,
        body: String,
        idToken: String
    ) async throws {
        let now = ISO8601DateFormatter().string(from: Date())
        _ = try await request(
            path: "documents/organizations/\(communityId)/memberPosts",
            method: "POST",
            idToken: idToken,
            body: ["fields": [
                "organizationId": stringValue(communityId),
                "memberUid": stringValue(userId),
                "memberName": stringValue(authorName),
                "title": stringValue(title),
                "body": stringValue(body),
                "imageURLs": arrayValue([]),
                "status": stringValue("submitted"),
                "adminReply": stringValue(""),
                "memberHasReadReply": boolValue(true),
                "createdAt": timestampValue(now),
                "updatedAt": timestampValue(now)
            ]]
        )
    }

    public func updateMemberPost(
        communityId: String,
        postId: String,
        title: String,
        body: String,
        idToken: String
    ) async throws {
        _ = try await request(
            path: "documents/organizations/\(communityId)/memberPosts/\(postId)"
                + "?updateMask.fieldPaths=title&updateMask.fieldPaths=body"
                + "&updateMask.fieldPaths=updatedAt",
            method: "PATCH",
            idToken: idToken,
            body: ["fields": [
                "title": stringValue(title),
                "body": stringValue(body),
                "updatedAt": timestampValue(ISO8601DateFormatter().string(from: Date()))
            ]]
        )
    }

    public func deleteMemberPost(
        communityId: String,
        postId: String,
        idToken: String
    ) async throws {
        _ = try await request(
            path: "documents/organizations/\(communityId)/memberPosts/\(postId)",
            method: "DELETE",
            idToken: idToken
        )
    }

    public func markReplyRead(
        communityId: String,
        postId: String,
        idToken: String
    ) async throws {
        _ = try await request(
            path: "documents/organizations/\(communityId)/memberPosts/\(postId)"
                + "?updateMask.fieldPaths=memberHasReadReply",
            method: "PATCH",
            idToken: idToken,
            body: ["fields": ["memberHasReadReply": boolValue(true)]]
        )
    }

    private func parseMemberPost(
        _ document: [String: Any],
        communityId: String
    ) -> MemberPost? {
        guard let fields = document["fields"] as? [String: Any],
              let name = document["name"] as? String,
              let id = name.split(separator: "/").last.map(String.init) else { return nil }
        let imageURLs = stringArray(fields, "imageURLs")
        var attachments = imageURLs.compactMap { rawURL -> PostAttachment? in
            guard let url = URL(string: rawURL) else { return nil }
            return PostAttachment(type: "image", name: "画像", url: url)
        }
        if let rawPDF = string(fields, "pdfURL"), let url = URL(string: rawPDF) {
            attachments.append(PostAttachment(type: "pdf", name: "PDF", url: url))
        }
        return MemberPost(
            id: id,
            communityId: string(fields, "organizationId") ?? communityId,
            authorUserId: string(fields, "memberUid")
                ?? string(fields, "authorUserId") ?? "",
            authorName: string(fields, "memberName")
                ?? string(fields, "authorName") ?? "会員",
            title: string(fields, "title") ?? "",
            body: string(fields, "body") ?? "",
            attachments: attachments,
            status: string(fields, "status") ?? "submitted",
            legacyAdminReply: string(fields, "adminReply"),
            memberHasReadReply: bool(fields, "memberHasReadReply") ?? true,
            createdAt: timestamp(fields, "createdAt"),
            updatedAt: timestamp(fields, "updatedAt")
        )
    }

    private func parsePublicPost(_ document: [String: Any]) -> PublicPost? {
        guard let fields = document["fields"] as? [String: Any],
              bool(fields, "isPublic") == true,
              let name = document["name"] as? String,
              let id = name.split(separator: "/").last.map(String.init) else { return nil }
        return PublicPost(
            id: id,
            authorUserId: string(fields, "authorUserId")
                ?? string(fields, "memberUid") ?? "",
            authorName: string(fields, "authorName")
                ?? string(fields, "memberName") ?? "会員",
            title: string(fields, "title") ?? "",
            categoryId: string(fields, "categoryId") ?? string(fields, "category"),
            body: string(fields, "body") ?? "",
            attachments: attachments(fields),
            createdAt: timestamp(fields, "createdAt")
        )
    }

    private func attachments(_ fields: [String: Any]) -> [PostAttachment] {
        guard let value = fields["attachments"] as? [String: Any],
              let array = value["arrayValue"] as? [String: Any],
              let values = array["values"] as? [[String: Any]] else { return [] }
        return values.compactMap { value in
            guard let map = value["mapValue"] as? [String: Any],
                  let attachmentFields = map["fields"] as? [String: Any],
                  let rawURL = string(attachmentFields, "url"),
                  let url = URL(string: rawURL) else { return nil }
            return PostAttachment(
                type: string(attachmentFields, "type") ?? "url",
                name: string(attachmentFields, "name") ?? "添付ファイル",
                url: url
            )
        }
    }

    private func request(
        path: String,
        method: String,
        idToken: String? = nil,
        body: [String: Any]? = nil
    ) async throws -> Any {
        guard !projectId.isEmpty,
              let url = URL(
                string: "https://firestore.googleapis.com/v1/projects/\(projectId)"
                    + "/databases/(default)/\(path)"
              ) else {
            throw URLError(.badURL)
        }
        var request = URLRequest(url: url)
        request.httpMethod = method
        if let idToken {
            request.setValue("Bearer \(idToken)", forHTTPHeaderField: "Authorization")
        }
        if let body {
            request.setValue("application/json", forHTTPHeaderField: "Content-Type")
            request.httpBody = try JSONSerialization.data(withJSONObject: body)
        }
        let (data, response) = try await session.data(for: request)
        guard let http = response as? HTTPURLResponse,
              (200..<300).contains(http.statusCode) else {
            throw URLError(.cannotParseResponse)
        }
        if data.isEmpty { return [:] }
        return try JSONSerialization.jsonObject(with: data)
    }

    private func string(_ fields: [String: Any], _ key: String) -> String? {
        (fields[key] as? [String: Any])?["stringValue"] as? String
    }

    private func bool(_ fields: [String: Any], _ key: String) -> Bool? {
        (fields[key] as? [String: Any])?["booleanValue"] as? Bool
    }

    private func stringArray(_ fields: [String: Any], _ key: String) -> [String] {
        guard let value = fields[key] as? [String: Any],
              let array = value["arrayValue"] as? [String: Any],
              let values = array["values"] as? [[String: Any]] else { return [] }
        return values.compactMap { $0["stringValue"] as? String }
    }

    private func timestamp(_ fields: [String: Any], _ key: String) -> Date? {
        guard let value = (fields[key] as? [String: Any])?["timestampValue"] as? String
        else { return nil }
        return ISO8601DateFormatter().date(from: value)
    }

    private func stringValue(_ value: String) -> [String: Any] {
        ["stringValue": value]
    }

    private func boolValue(_ value: Bool) -> [String: Any] {
        ["booleanValue": value]
    }

    private func timestampValue(_ value: String) -> [String: Any] {
        ["timestampValue": value]
    }

    private func arrayValue(_ values: [[String: Any]]) -> [String: Any] {
        ["arrayValue": ["values": values]]
    }
}
